### ISSUE TITLE
Go divide by zero behaviour

### ISSUE THREAD

**Issue by elliott5** [go2hx/go2hx#31]
In Go, all integers panic when divided by zero. This is a different behaviour from Haxe.

For floats & complexes, NaN should be returned. This means the code for complex number division in TardisGo was wrong (and so is wrong now in go2hx).

Go has a specific test for divide-by-zero behaviour at https://golang.org/test/zerodivide.go

**Comment by PXshadow** [comment]
I'm currently using the same method tinygo is of returning null for recover, a decent porition of the test uses recover however, so it seems an implementation of recover is needed to pass this test.

**Comment by elliott5** [comment]
Prototype implementation of recover():
- I took the example in the go blog https://blog.golang.org/defer-panic-and-recover. 
- Used go2hx to translate it into Haxe. 
- Then amended the generated Haxe to execute properly using Haxe `try{}catch(e){}`, with `NOTE` comments to show what is changed. 
- My prototype code is at: https://gist.github.com/elliott5/0118b29073180ddf327c4d7166c21ec7

**Comment by PXshadow** [comment]
You are a genius well done, the exception handling is suberp my only question is, do you have a method to calculate where the try block should start and end? Reading the code it can be figured out but if there is some pattern for the compiler to use when typing.

**Comment by elliott5** [comment]
Because the only place that `recover()` can be called and get a non-nil result is in a `defer` function, I think the scope of the `try` block has to be the body of the function. 

Since Go requires that all the `defer` functions be run where there is a panic, where there are `defer` functions a `try{}catch{}` will also be required - as per the prototype function `g()` where the `defer` functions run after a panic, even though the panic is not `recover()`ed in them.

**Comment by PXshadow** [comment]
Defer calls for functions that use rest arguments is currently blocked by a Haxe issue as far as I'm aware https://github.com/HaxeFoundation/haxe/issues/10196 , I can't think of any decent workaround as the arguments for defer by the go spec need to be declared where the defer keyword is and then the function ran at the bottom, therefore it is runtime functionality.

**Comment by elliott5** [comment]
Workaround prototype at https://try.haxe.org/#7E7EB22e

**Comment by elliott5** [comment]
I don't think every function that uses `defer` should begin with `Go.recover_exception = null;`, because if a `panic` is currently in progress (possible if a `defer`ed function calls another function which uses `defer`), this would loose the information about the current Haxe exception.

**Comment by PXshadow** [comment]
> Workaround prototype at https://try.haxe.org/#7E7EB22e

Thank you this is very useful, a little silly how straight forward it is and yet I totally missed it. I will use this prototype for implementation.

**Comment by elliott5** [comment]
For build 6698ff8...
Function `f()`  `_defercall_0` generated Haxe:
```
private function f():Void {
    var recover_exception:Dynamic = null;
    var deferstack:Array<Array<Dynamic>> = [];
    var _defercall_0 = null;
    try {
        _defercall_0 = function() {
            var a = function() {
                {
                    var r = recover();
                    if (!Go.isNull(r)) {
                        println((("Recovered in f" : GoString)), r);
                    };
                };
                return;
            };
            a();
        };
```
Looking at the Haxe Dump file (after macro expansion), is being expanded to:
```
	static function f() {
		var recover_exception = [null];
		var _defercall_0 = null;
		try {
			_defercall_0 = function() {
				var a = function() {
					var r = recover_exception[0];
					recover_exception[0] = null;
					var r = r;
				};
				a();
			};
```
...which obviously never prints "Recovered in f".

My solution, make `recover_exception` global, and simplify `recover()`:
```
var recover_exception:Dynamic = null; // needs to be global
inline function erecover():Dynamic { // test replacement for recover()
    var __r = recover_exception;
    recover_exception=null;
    return __r;
}
```
Also swap `!Go.isNull(r)` (which is never `true`) to `r!=null`.

With those changes, everything works properly using `haxe --interp`.

... but does not work properly with the `hl` target. 

In the `hl` target, in the `catch` block `_defercall_0` is always `null` even though it was set in the `try` block. This seems like another `Haxe/HL` issue to me, but who knows. 

My work-around is to create a global `var _defer_stack:Array<Void->Void> = [];`.
To create a deferred function `_defer_stack.push(function():Void { /* function body */});`.
To run a deferred function `_defer_stack.pop()();`.
At the start of each function create a new local variable `var _start_defer_stack_length = _defer_stack.length;`
Then to unwind the defer stack for a particular function:
```
            while(_defer_stack.length>_start_defer_stack_length)
                _defer_stack.pop()();
```

This approach works for both `hl` and `--interp` targets.

My modified Haxe is at https://gist.github.com/elliott5/cfe0995875293f985cce5238c7902af3

Note that I had to rewrite Go `panic(fmt.Sprintf("%v", i))` to Haxe `throw Std.string(i);` because the `"%v"` seems to cause an unrelated panic.

**Comment by elliott5** [comment]
To be correct, the Go `recover()` builtin returns an `interface{}` which is modelled in go2hx as a Haxe `Any` value.

So the code should actually be something like:
```
inline function recover():Any {
	var __r:Any = recover_exception;
	recover_exception = null;
	return __r;
}
```

**Comment by PXshadow** [comment]
This is very useful thank you! I sent in an issue to Haxe, I'm fine using the other targets for now. I'm not sure why recover_exception is required to be a global, It seems to be working now scopped within the function, I'm fine making it global does this at all relate to your comment with Go.recover_exception though? "Go.recover_exception = null;, because if a panic is currently in progress (possible if a defered function calls another function which uses defer), this would loose the information about the current Haxe exception." If the vital diffrence is the fact that the recover_exception is not being set as null for each function, is there no possibility of an inprogress panic messing the logic up of another, as I believe in the future these could also be in goroutines?

**Comment by elliott5** [comment]
Yes, exactly so @PXshadow, the exception must not be lost before it can be recovered.

**Comment by PXshadow** [comment]
@elliott5 I think I've ran into a brick wall any idea how to support a defer like this?
```go
func testDeferLoop() {
	for j := 0; j < 4; j++ {
		defer deferred("loop", j)
	}
}```

It seems the current generation approach can't solve for this completly at compile time. Perhaps a relook at the rest argument system and see if it's possible to add it into the function args at runtime?

**Comment by elliott5** [comment]
You never know at compile time how many `defer` statements will be executed by the logic of a function.

The best thing to do is to use a stack to hold the defer calls, then you can unwind the stack when the function exits (in whatever way). 

I outlined an example approach to this 4 days ago above.

**Comment by PXshadow** [comment]
Oh yes your method works wonderfully shoot I didn't see the read the unwinding part my apologies! When I wrote you the issue I had an idea from it and spent the last 30 mins working on a macro fix the unwinding method seems more clean straight forward though, here is the code.

```haxe
import haxe.macro.Context;
import haxe.macro.Expr;
function main() {
	var deferstack:Array<{rest:Int,args:Array<Dynamic>,func:haxe.Constraints.Function}> = [];
	deferstack.push(callMethod(test,1,2,3,4,5,6,7,8,9,10));

	for (defer in deferstack) {
		if (defer.rest == -1) {
			Reflect.callMethod(null,defer.func,defer.args);
			continue;
		}
		var args = defer.args.slice(0,defer.rest);
		if (args.length != defer.args.length) {
			var rest = defer.args.slice(defer.rest);
			args.push(rest);
		}
		Reflect.callMethod(null,defer.func,args);
	}
	trace(deferstack);
}

function test(x:Int,y:Int,...z:Int) {
	trace('x: $x y: $y z: $z');
}


macro function callMethod(func:Expr,args:Array<Expr>) {
	var restIndex = -1;
	var type = Context.follow(Context.typeof(func));
	switch type {
		case TFun(args, ret):
			var last = args.pop();
			if (last != null && !last.opt) {
				switch last.t {
					case TAbstract(t, params):
						var t = t.get();
						if (t.name == "Rest" && t.pack.length == 1 && t.pack[0] == "haxe")
							restIndex = args.length;
					default:
				}
			}
		default:
	}
	return macro {
		func: $func,
		args: $a{args},
		rest: $v{restIndex},
	};
}```

**Comment by elliott5** [comment]
In the code above the order of execution using `for (defer in deferstack) {` is wrong - the stack entries need to be `pop()`ed, see https://tour.golang.org/flowcontrol/13

Using `Reflect.callMethod(null,defer.func,args);` with the `args` all held in `Array<Dynamic>`  will  be much slower at runtime than using a `Void-Void` closure. 

You will also need code in addition to `callMethod()` to deal with a Go `defer` that is a closure:
https://stackoverflow.com/questions/16010694/how-golangs-defer-capture-closures-parameter

As you can see from the `stackoverflow` post above, even using a `Void->Void` stack, there is an issue around capturing the correct value of the variables captured by the closure, so there may still be a role for a `args` stack to go alongside the `Void->Void` function stack - although I would always prefer it to be typed (rather than `Dynamic`) as this will produce faster code.

**Comment by PXshadow** [comment]
Thank you this advice is very helpful @elliott5 the closure but will defiently be tricky I've ran into it a little bit, I'll run some tests and also get defer functioning, the last part of defer that it a little bit tricky is dealing with returns as a try catch makes it so the return might not get called within try and therefore will fail with "expected return"  I've removed most of these errors in the code simply by requiring recover to be present in order to generate a try catch but it's not a fix to the problem.  I assume I can just add the default return value of the function, here is generated code as an example. Thanks again for the expertise :)

```haxe
/**
    func error_(fn func()) (error string) {
    	defer func() {
    		if e := recover(); e != nil {
    			error = e.(runtime.Error).Error()
    		}
    	}()
    	fn()
    	return ""
    }
**/
private function error_(fn:() -> Void):GoString {
    var recover_exception:Dynamic = null;
    var deferstack:Array<Array<Dynamic>> = [];
    var _defercall_0 = null;
    try {
        _defercall_0 = function() {
            var a = function() {
                var error:GoString = "";
                {
                    var e = recover();
                    if (!Go.isNull(e)) {
                        var error:GoString = "";
                        error = ((e : stdgo.runtime.Runtime.Error)).error();
                    };
                };
                return error;
            };
            a();
        };
        fn();
        {
            {
                _defercall_0();
            };
            return (("" : GoString));
        };
    } catch(e) {
        recover_exception = e;
        if (recover_exception != null) throw recover_exception;
    };
}
```

**Comment by elliott5** [comment]
The generated Haxe is wrong because there is no 
```
            {
                _defercall_0();
            };
```
block within the `catch` branch. 

Unwinding the `defer` stack needs to happen in multiple places - after every explicit `return` and after the end of a function, as well as in the `catch` branch.

But the actual return of the function needs to happen after the `try{}catch{}` blocks, so that the `defer` closure can set the return value, as in this example.

So I think the generated Haxe should be of the form below. Please note the 3 ADD & 4 REMOVE commented lines:
```
private function error_(fn:() -> Void):GoString {
    var error:GoString = "";  // ADD this is the return value, so needs to be visible across both try & catch branches
    var recover_exception:Dynamic = null;
    var deferstack:Array<Array<Dynamic>> = [];
    var _defercall_0 = null;
    try {
        _defercall_0 = function() {
            var a = function() {
                // var error:GoString = ""; // REMOVE - error needs to be captured by this closure from the error_() context
                {
                    var e = recover();
                    if (!Go.isNull(e)) {
                        // var error:GoString = ""; // REMOVE - error needs to be captured by this closure from the error_() context
                        error = ((e : stdgo.runtime.Runtime.Error)).error();
                    };
                };
                // return error; // REMOVE - nothing is returned here
            };
            a();
        };
        fn();
        {
            {
                _defercall_0();
            };
            // return (("" : GoString)); // REMOVE - error was set to "" on declaration at the start of the function
        };
    } catch(e) {
        recover_exception = e;
            {
                _defercall_0(); // ADD - unwind the defer stack, catching recover_exception into error in this case
            };
        if (recover_exception != null) throw recover_exception;
    };
    return error; // ADD - return the named return value
}
```

**Comment by elliott5** [comment]
Looking at the code above, it may be that the body of the `try` block needs to be a closure too, so that the `return` control flow works properly, of the form:
```
function x():Int {
     var _return_value:Int=0;
     //... other variable declarations at function level, including for defer
    var _body = function() {
         // ...whatever the body of the function does, including pushing defer functions, possibly with multiple return points
    };
    try {
       _body();
       _defercall_0(); // or whatever method is used to unwind the defer stack
   } catch (e) {
         recover_exception = e;
         _defercall_0(); // or whatever method is used to unwind the defer stack
         if (recover_exception != null) throw recover_exception;
    };
    return _return_value;
}

**Comment by elliott5** [comment]
Go code below for reference of the beautiful example discussed in the posts above (named returned value and defer and try/catch), not working yet as at bd286f62196163840fee01550e2bef0a8c45752e.
```
package main

import (
	"fmt"
	"runtime"
)

func error_(fn func()) (error string) {
	defer func() {
		if e := recover(); e != nil {
			error = e.(runtime.Error).Error()
		}
	}()
	fn()
	return ""
}

func main() {
	z := 0
	make_error := func() {
		z = 42 / z
	}
	fmt.Println("error_()=>", error_(make_error))
	// output:
	// error_()=> runtime error: integer divide by zero
}
```

It may be worth re-naming this issue, to make getting this code working the main focus.

**Comment by PXshadow** [comment]
> It may be worth re-naming this issue, to make getting this code working the main focus.

Yeah good idea, sorry for the delay, so many tests to run I didn't manage to run and see that divide by zero test working as well as something similar as you have above, I'll work on it right away @elliott5 .

**Comment by elliott5** [comment]
Well done for getting the `error_()=> runtime error: integer divide by zero` example working @PXshadow !

**Comment by PXshadow** [comment]
@elliott5 Thank you so much! Was really nice to manage to get it passing, I switched it from just a plain interface Error to a runtime Error in order to allow it to cast, that was the big hurdle.

**Comment by PXshadow** [comment]
I believe the divide by zero behavior is integrated I'm not 100% sure though, let me know if there is anything missing @elliott5 if not feel free to close.

**Comment by PXshadow** [comment]
Confirmed implemented as multiple std Math tests are now passing.
